common/src/types/v0/store/node.rs

```diff
@@ -1,95 +1,318 @@
+    /// Endpoint of the io-engine instance (gRPC).
     endpoint: String,
     /// Node labels.
     labels: NodeLabels,
-    /// Cordon labels.
+    /// Cordon/drain state.
+    #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(default)] // Ensure backwards compatibility in etcd when upgrading.
-    cordon_labels: Vec<String>,
+    cordon_drain_state: Option<CordonDrainState>,
 }
 
 impl NodeSpec {
-    /// Return a new `Self`
+    /// Return a new `Self`.
     pub fn new(
         id: NodeId,
         endpoint: String,
         labels: NodeLabels,
-        cordon_label: Option<Vec<String>>,
+        cordon_drain_state: Option<CordonDrainState>,
     ) -> Self {
         Self {
             id,
             endpoint,
             labels,
-            cordon_labels: cordon_label.unwrap_or_default(),
+            cordon_drain_state,
         }
     }
-    /// Node identification
+    /// Node identification.
     pub fn id(&self) -> &NodeId {
         &self.id
     }
-    /// Node gRPC endpoint
+    /// Node gRPC endpoint.
     pub fn endpoint(&self) -> &str {
         &self.endpoint
     }
-    /// Node labels
+    /// Node labels.
     pub fn labels(&self) -> &NodeLabels {
         &self.labels
     }
-    /// Node gRPC endpoint
+    /// Node labels.
+    pub fn cordon_drain_state(&self) -> &Option<CordonDrainState> {
+        &self.cordon_drain_state
+    }
+
+    /// Node gRPC endpoint.
     pub fn set_endpoint(&mut self, endpoint: String) {
         self.endpoint = endpoint
     }
+
     /// Cordon node by applying the label.
     pub fn cordon(&mut self, label: String) {
-        self.cordon_labels.push(label);
+        // the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
+                    // just add the label
+                    state.add_cordon_label(&label);
+                }
+                CordonDrainState::Draining(state) => {
+                    // just add the label
+                    state.add_cordon_label(&label);
+                }
+                CordonDrainState::Drained(state) => {
+                    // just add the label
+                    state.add_cordon_label(&label);
+                }
+            },
+            None => {
+                //add the label and set the state to cordoned
+                self.cordon_drain_state =
+                    Some(CordonDrainState::Cordoned(CordonedState::new(vec![label])));
+            }
+        }
+    }
+
+    /// Drain node by applying the drain label.
+    pub fn set_drain(&mut self, label: String) {
+        // the the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
+                    // set state to draining and add label
+                    let drainlabels = vec![label];
```

Comment:
what about moving this logic to a function drain in CordonState, eg:
```rust
impl CordonState {
  fn into_draining(self, drainlabel: &str) -> DrainState {
     DrainState::new(self.cordonlabels, vec![drainlabel])
  }
}
...
self.cordon_drain_state = state.into_draining(label);
```

Reply:
fixed